gen/enum.go

```diff
@@ -101,6 +106,96 @@ func enum(g Generator, spec *compile.EnumSpec) error {
 			<end>
 			return fmt.Sprintf("<$enumName>(%d)", <$w>)
 		}
+
+		func (<$v> <$enumName>) MarshalJSON() ([]byte, error) {
+			<if len .Spec.Items>
+				switch int32(<$v>) {
+				<range .UniqueItems>
+					case <.Value>:
+						return ([]byte)("\"<.Name>\""), nil
+				<end>
+				}
+			<end>
+			return ([]byte)(<$strconv>.FormatInt(int64(<$v>), 10)), nil
+		}
+
+		<$text := newVar "text">
+		func (<$v> *<$enumName>) UnmarshalJSON(<$text> []byte) error {
+			<$d := newVar "d">
+			<$t := newVar "t">
+
+			<$d> := <$json>.NewDecoder(<$bytes>.NewReader(<$text>))
+			<$d>.UseNumber()
+			<$t>, err := <$d>.Token()
+			if err != nil {
+				return err
+			}
+
+			<$ok := newVar "ok">
+			if <$w>, <$ok> := <$t>.(<$json>.Number); <$ok> {
+				<$x := newVar "x">
+				<$x>, err := <$w>.Int64()
+				if err != nil {
+					return err
+				}
+				if <$x> <">="> 0x80000000 {
+					return <$fmt>.Errorf("enum overflow from JSON %q for %q", <$text>, "<$enumName>")
+				}
+				if <$x> <"<"> -0x80000000 {
+					return <$fmt>.Errorf("enum underflow from JSON %q for %q", <$text>, "<$enumName>")
+				}
+				*<$v> = (<$enumName>)(<$x>)
+				return nil
+			}
+
+			if <$w>, <$ok> := <$t>.(string); <$ok> {
+				switch string(<$w>) {
+				<$enum := .Spec>
+				<range .Spec.Items>
+					case "<.Name>":
```

Comment:
right now we enforce enum items uniqueness case insensitively (`enum { x X }` is not allowed). so yes, in theory we could safely parse enum item names as case insensitive.

for now, let's keep things simple, and match as it is in the thrift file. we can always decide later to relax one of the two rule.